File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int helper(TreeNode* root, int &diameter){
        if(root == NULL) return 0;

        int lh = helper(root -> left, diameter);
        int rh = helper(root -> right, diameter);

        diameter = max(diameter, lh+rh);
        return max(lh, rh) + 1;

    }

    int diameterOfBinaryTree(TreeNode* root) {
        int diameter = 0;
        helper(root, diameter);
        return diameter;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = [0]
        
        def findDiameter(root)->int:
            if root is None:
                return 0

            lh = findDiameter(root.left)
            rh = findDiameter(root.right)

            diameter[0] = max(diameter[0] , lh+rh)
            return max(lh,rh) + 1

        findDiameter(root)
        return diameter[0]